### ISSUE TITLE
Kernel security features

### ISSUE THREAD

**Issue by p4zuu** [coconut-svsm/svsm#448]
I think we can start tracking down the security features we could add to the COCONUT kernel to improve its security:

- [ ] KASLR
- [ ] Read-only GDT and IDT
- [ ] SMEP and SMAP
- [ ] Heap hardening

This list aims to be as complete as possible, so please comment or reach me if you want to add or modify elements from the list, or would like to assign yourself to a task :)

**Comment by Freax13** [comment]
Shadow stacks :)

**Comment by 00xc** [comment]
We could look into LLVM's [KCFI](https://doc.rust-lang.org/stable/unstable-book/compiler-flags/sanitizer.html#kernelcontrolflowintegrity).

Also we could perhaps consider a KPTI-like mechanism depending on our threat model regarding userspace.

**Comment by Freax13** [comment]
Rust has support for [KASAN](https://doc.rust-lang.org/stable/unstable-book/compiler-flags/sanitizer.html#kerneladdresssanitizer).

**Comment by p4zuu** [comment]
> Rust has (unstable) support for [KASAN](https://doc.rust-lang.org/stable/unstable-book/compiler-flags/sanitizer.html#kerneladdresssanitizer).

KASAN would be very useful in testing env indeed, I'll track it here too. @00xc already had a look into KASAN but struggled a bit with the linker as far as I remember. In any cases, we would still have to write the KASAN handlers ourselves, right?

**Comment by Freax13** [comment]
I have a working implementation of KASAN in [mushroom](https://github.com/Freax13/mushroom/blob/main/tee/kernel/src/sanitize.rs).

> In any cases, we would still have to write the KASAN handlers ourselves, right?

Yes, there's an interface that has to implemented by the kernel.

**Comment by 00xc** [comment]
Other than the linking issues I had some boot hangs. I managed to get a working build by disabling stack instrumentation. I did not start the actual implementation though.

**Comment by Freax13** [comment]
We could extend our elf loader to support RELRO.

**Comment by Freax13** [comment]
I'd like to take a stab at shadow stacks.

**Comment by nagyation** [comment]
I am interested in implementing "RELRO support for the ELF loader", Have this already been discussed somewhere or assigned to someone already?

**Comment by 00xc** [comment]
> I am interested in implementing "RELRO support for the ELF loader", Have this already been discussed somewhere or assigned to someone already?

I don't think so. Feel free to join the community call on Wednesday and bring up the topic. I'm sure others will have comments.
